### PR TITLE
Parse Binary and Clock variables

### PR DESCRIPTION
**Pull request by filip-stenstrom**
- Adds basic parsing of Clock and Binary variables (no/few error checks - and not tested)
- Does NOT handle their TypeDefinitions

### CODE REVIEW

File: Config.cmake/runtime_test.cmake [hunk 1]
```diff
@@ -11,79 +11,54 @@
 #    You should have received a copy of the FMILIB_License.txt file
 #    along with this program. If not, contact Modelon AB <http://www.modelon.com>.
 
-set(RTTESTDIR ${FMILIBRARYHOME}/Test)
+set(FMIL_TEST_DIR ${FMILIBRARYHOME}/Test)
+set(CATCH_INCLUDE_DIR ${FMILIB_THIRDPARTYLIBS}/Catch2/single_include/catch2)
+set(TEST_OUTPUT_FOLDER ${CMAKE_CURRENT_BINARY_DIR}/Testing)
+set(CATCH_INCLUDE_DIR "${FMILIB_THIRDPARTYLIBS}/Catch2/single_include/catch2/")
```
Comment (filip-stenstrom) [inline Config.cmake/runtime_test.cmake:17]: Mainly moved to top of file for readability, and renamed RTTESTDIR -> FMIL_TEST_DIR.

File: Config.cmake/runtime_test.cmake [hunk 2]
```diff
@@ -11,79 +11,54 @@
...
 
-add_library(fmi_testutil STATIC ${RTTESTDIR}/fmi_testutil.c)
+include_directories(
+    ${FMIL_TEST_DIR}
+    ${CATCH_INCLUDE_DIR}
+)
```
Comment (filip-stenstrom) [inline Config.cmake/runtime_test.cmake:22]: To avoid repeating everywhere

File: Config.cmake/runtime_test.cmake [hunk 3]
```diff
@@ -11,79 +11,54 @@
...
-    ${RTTESTDIR}/FMI3/fmi3_basic_capi_test.cpp)
-target_include_directories(fmi3_basic_capi_test
-    PUBLIC
-    ${CATCH2_INCLUDE_DIR}
-    ${FMIIMPORTDIR}/src)
-target_link_libraries(fmi3_basic_capi_test
-    PUBLIC
-    Catch
-    ${FMILIBFORTEST})
-add_test(ctest_fmi3_basic_capi_test_cpp fmi3_basic_capi_test)
```
Comment (filip-stenstrom) [inline Config.cmake/runtime_test.cmake:60]: Moved to the fmi3 test file

File: Config.cmake/runtime_test.cmake [hunk 4]
```diff
@@ -11,79 +11,54 @@
...
-add_executable(jm_vector_test_cpp
-    ${RTTESTDIR}/jm_vector_test.cpp)
-set_source_files_properties(${RTTESTDIR}/jm_vector_test.cpp PROPERTIES LANGUAGE CXX)
-target_include_directories(jm_vector_test_cpp
-    PUBLIC
-    ${CATCH2_INCLUDE_DIR})
-target_link_libraries(jm_vector_test_cpp
-    Catch
-    ${JMUTIL_LIBRARIES})
-add_test(ctest_jm_vector_test_cpp jm_vector_test_cpp)
```
Comment (filip-stenstrom) [inline Config.cmake/runtime_test.cmake:47]: Deleted - it was just for testing catch, which we now have in other tests

File: Test/FMI3/fmi3_import_variable_test.c [hunk 5]
```diff
@@ -1,128 +0,0 @@
-#include <stdio.h>
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_variable_test.c:1]: Whole file converted to Catch2, i.e. is now a .cpp file

File: dev/diagrams/diagrams.drawio [hunk 6]
```diff
@@ -1 +1,647 @@
...l7jAHILt5Rbf99yM0vVycGqZZAzTKVoeb7VQEOswcglVhT6a9QoPTZEqV/t9dwJO9qulChyQB1SDZ8K3YbGSieyBwhOqBedKguC5voUJn2+EWVPlKrUTykz0qxXV+J+zB7apaASf7nrR9b/4MPpv/9eOdRtYFzM54i/phMIyTFNtmkD//L9Kfiy3i/Cd+fL8SdoFy8ObXtuD8Mt24P/IdpDcLzGo7doDyed0hJ0yKNriZtrjgT+A2bGQOfNHk1ZD8beYKorR1l6prHKpKoeujpwidUX1al9kGEIDfEImCEyBvHvxFBwwgjjZt9obv5epMUm2neUW38X2jf22p9eLynaPxpFw76TjKoG70aBTo6rdmy9Dt31bD4Xiqjr8t9DmkdbpPl/jdjHCNlQKBi5ya7H+MOWkXDzlLmrvVt2mtWH5+zwvaYexS2B+IMqGtrfg/xU00gQrHWR0XtPI0HQSSSWX4QRAiDEcNuEAuukqpIR+oVA+jLFwkSYF7EsuIYwFRIBqK7Y3K7TTjz9S1t+dHTUHSWX6ap2OLOjrI8WSDHkfD+dTlU5URa8oNX4DPW6bU/qSqkJHKvR+Kcpm6q0HHA47a+p+4XhtB5g23SIdqeDum1+yCyrQPNzcHT2Fw==</diagram><diagram id="6je8l0Wj5OvteJp03YQ2" name="callbacks">rZPBboMwDIafJsdKkEC160q7Vpq2SzdttyktbsgUCApuS/v0C2AKqJdNmpSD88Wxld9/mEjyeu1kmb3YFAzjQVozsWSchxHnrFlBeunIQw+U0yklDWCrr0AwIHrUKVSTRLTWoC6ncG+LAvY4YdI5e56mHayZdi2lgjuw3UtzTz90ihm9Ig4GvgGtsr5zGNBJLvtkAlUmU3seIbFiInHWYhfldQKmEa/XBXn4peX3vBLvm+T58zWqV9GsK/b0lyu3Jzgo8H9L0yxP0hxJrxxy6y6Mz41vtdg5H6kmMlYpcPfczw2hRpIIL73uLRSLDHPjQehDB5W+yl2bEPh9aXWB7XTjBYuXnsgj2qpzUHNBGq0KHxs4NKVO4FD7wT4SRlt6WpVyrwv11myWs8iTX6pFqjZVoR55hdRbg80BGyUCOuUxqUU/QfROOQ++mhPKRpaKiElysrpVHqblAxpYvx181Z6NfqdY/QA=</diagram></mxfile>
```
Comment (filip-stenstrom) [inline dev/diagrams/diagrams.drawio:1]: This is a file containing some UML. I updated some of it, and it seems a new file format is being used why everything is changed.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 7]
```diff
@@ -377,9 +392,98 @@ int fmi3_xml_set_attr_enum(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu_
...
-int fmi3_xml_set_attr_boolean(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu_t elmID, fmi3_xml_attr_enu_t attrID, int required, unsigned int* field, unsigned int defaultVal) {
+int fmi3_xml_set_attr_boolean(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu_t elmID, fmi3_xml_attr_enu_t attrID,
+        int required, unsigned int* field, unsigned int defaultVal)
+{
     jm_name_ID_map_t fmi_boolean_i_dMap[] = {{"true", 1},{"false", 0}, {"1", 1},{"0", 0}, {0,0}};
-    return fmi3_xml_set_attr_enum(context,elmID, attrID,required, field, defaultVal, fmi_boolean_i_dMap);
+    return fmi3_xml_set_attr_enum(context, elmID, attrID, required, field, defaultVal, fmi_boolean_i_dMap);
+}
+
+// TODO: For FMI3, do we want to use bool in the getters for boolean attributes, or keep using unsigned int?
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_parser.c:402]: That would mean that we would need to change the return value of some API, but also make code more consistent.

I mean, it would be good if we could change all boolean functions to return `bool`. Right now we mainly return `int` but at least the one for getting FMU capabilities return `unsigned int`, probably because that's how they were parsed with this fcn.

We prob need a meeting to discuss it but early input is welcome

File: src/XML/src/FMI3/fmi3_xml_type_impl.h [hunk 8]
```diff
@@ -26,47 +26,119 @@
...
+    fmi3_uint8_t*   array8u;
+    fmi3_int64_t    scalar64s;
+    fmi3_int32_t    scalar32s;
+    fmi3_int16_t    scalar16s;
+    fmi3_int8_t     scalar8s;
+    fmi3_uint64_t   scalar64u;
+    fmi3_uint32_t   scalar32u;
+    fmi3_uint16_t   scalar16u;
+    fmi3_uint8_t    scalar8u;
+} fmi3_int_union_t;
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_type_impl.h:55]: Only moved

File: src/XML/src/FMI3/fmi3_xml_type_impl.h [hunk 9]
```diff
@@ -137,23 +188,6 @@ typedef struct fmi3_xml_int_type_props_t {
...
-    char itemDesciption[1];
-} fmi3_xml_enum_type_item_t;
-
-static int fmi1_xml_compare_enum_val (const void* first, const void* second) {
-    const jm_named_ptr* a = first;
-    const jm_named_ptr* b = second;
-    fmi3_xml_enum_type_item_t* ai = a->ptr;
-    fmi3_xml_enum_type_item_t* bi = b->ptr;
-    return (ai->value - bi->value);
-}
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_type_impl.h:155]: Renamed to fmi3 and moved

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 10]
```diff
@@ -605,49 +686,160 @@ int fmi3_xml_get_enum_variable_max(fmi3_xml_enum_variable_t* v){
...
 }
 
-const char* fmi3_xml_get_string_variable_start(fmi3_xml_string_variable_t* v){
+int fmi3_xml_get_enum_variable_start(fmi3_xml_enum_variable_t* v) {
     fmi3_xml_variable_t* vv = (fmi3_xml_variable_t*)v;
     if(fmi3_xml_get_variable_has_start(vv)) {
-        fmi3_xml_variable_start_string_t* start = (fmi3_xml_variable_start_string_t*)(vv->type);
-        return start->start;
+        fmi3_xml_int_variable_start_t* start = (fmi3_xml_int_variable_start_t*)(vv->type);
+        return start->start.scalar32s;
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_variable.c:693]: Moved code (yes, diff looks strange)

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 11]
```diff
@@ -686,34 +878,196 @@ void fmi3_xml_variable_free_internals(jm_callbacks* callbacks, fmi3_xml_variable
     }
 }
 
-int fmi3_xml_handle_Variable(fmi3_xml_parser_context_t *context, const char* data) {
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_variable.c:689]: I moved out some long functions for getting variability etc. attributes, and that makes the diff strange. I suggest to look at the actual code before trying to read this diff.

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 12]
```diff
@@ -746,107 +1100,25 @@ int fmi3_xml_handle_Variable(fmi3_xml_parser_context_t *context, const char* dat
...
-
-            if (variable->variability != fmi3_causality_enu_input && !multipleSet) {
-                fmi3_xml_parse_error(context, "Only variables with causality='input' can have canHandleMultipleSetPerTimeInstant=false");
-                return -1;
-            }
-        }
+        if (fmi3_xml_variable_process_attr_causality_variability_initial(context, variable, elm_id)) return -1;
+        if (fmi3_xml_variable_process_attr_previous(context, variable, elm_id))    return -1;
+        if (fmi3_xml_variable_process_attr_multipleset(context, variable, elm_id)) return -1;
+        if (fmi3_xml_variable_process_attr_clocks(context, variable, elm_id)) return -1;
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_variable.c:1121]: This is the actual new thing I added here. Rest is just refactor.

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 13]
```diff
@@ -1511,12 +2017,32 @@ int fmi3_xml_handle_ModelVariables(fmi3_xml_parser_context_t *context, const cha
...
+            size_t size = jm_vector_get_size(jm_named_ptr)(&md->variablesByName);
+            md->variablesByVR = jm_vector_alloc(jm_voidp)(size, size, md->callbacks);
+            if (!md->variablesByVR) {
+                fmi3_xml_parse_fatal(context, "Could not allocate memory");
+                return -1;
+            }
+            for (size_t i = 0; i < size; ++i) {
+                jm_vector_set_item(jm_voidp)(md->variablesByVR, i, jm_vector_get_item(jm_named_ptr)(&md->variablesByName, i).ptr);
             }
+            jm_vector_qsort(jm_voidp)(md->variablesByVR, fmi3_xml_compare_vr_and_original_index);
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_variable.c:2045]: This code is mainly shuffled around a bit. The purpose is to create the variablesByVR list before we resolve the previous variable.

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 14]
```diff
@@ -1322,7 +1742,91 @@ int fmi3_xml_handle_StringVariable(fmi3_xml_parser_context_t *context, const cha
...
+        // TODO: Or just warning and pad with trailing '0'?
+        return -1;
+    }
+
+    const char* pos = hexstr;
+    size_t nByte = len / 2;
+    for (size_t i = 0; i < nByte; i++) {
+        for (size_t j = 0; j <= 1; j++) {
+            char ch = pos[j];
+            if (!((ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'f') || (ch >= 'A' && ch <= 'F'))) {
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_variable.c:1775]: Would be sensible/worthwhile to only do this type of evaluation if sscanf fails?

**Review by PeterMeisrimelModelon — COMMENTED** [review]
Looks very good!

Some overall comments not bound to any specific code segment:

1. Comment style // vs /* */ is a bit inconsistent. Is this worthwhile fixing? Does this need to be C89 compliant?
2. There are quite a few TODOs, XXX, and 1 BUG in comments. Are we keeping proper track of all of these, are theses expected to be resolved within this MR?
3. There are a quite a few "User is responsible for freeing memory etc comments" I think it could be useful provide the how, since it might not be clear how to <properly> do this. And even if it is a simple free(...), this may not be obvious either.

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 15]
```diff
@@ -377,9 +392,98 @@ int fmi3_xml_set_attr_enum(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu_
...
+        char ch = *cur;
+
+        while ((ch == ' ') || (ch == '\t') || (ch == '\n') || (ch == '\r')) {
+            ch = *(++cur);
+            if (!ch) break;
+        }
+        if (!ch) break;
+
+        if (sscanf(cur, "%" SCNu32, &vr) != 1) {
+            // TODO: Improve error
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c:469]: fixed

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 16]
```diff
@@ -377,9 +392,98 @@ int fmi3_xml_set_attr_enum(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu_
...
+
+    if (fmi3_xml_get_attr_str(context, elmID, attrID, required, &strVal)) {
+        return -1;
+    }
+
+    if (!strVal && !required) {
+        *field = defaultVal;
+        return 0;
+    }
+
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c:416]: `fmi3_xml_get_attr_str` will return false in that case, so wouldn't even get here. It would probably make more sense to remove the check for `required`

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 7 above)
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c:402]: Okay good. I'll bring it up later for more discussion

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 17]
```diff
@@ -115,13 +118,20 @@ const char *fmi3_xmlAttrNames[fmi3_xml_attr_number] = {
...
 #define fmi3_xml_scheme_EnumerationVariable  {fmi3_xml_elmID_Variable,   fmi3_xml_elmID_ModelVariables,      0,       1}
 #define fmi3_xml_scheme_Dimension            {fmi3_xml_elmID_none,       fmi3_xml_elmID_Variable,            0,       1}
 
+#define fmi3_xml_scheme_BinaryVariableStart  {fmi3_xml_elmID_none,       fmi3_xml_elmID_BinaryVariable,      0,       1}
+
 #define fmi3_xml_scheme_Annotations          {fmi3_xml_elmID_none,       fmi3_xml_elmID_Variable,            1,       0}
 #define fmi3_xml_scheme_VariableTool         {fmi3_xml_elmID_none,       fmi3_xml_elmID_Annotations,         0,       1}
 
+// Not used except for setting up the element handler framework:
+#define fmi3_xml_scheme_Start                {fmi3_xml_elmID_none,       fmi3_xml_elmID_none,                0,       0}
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c:133]: There is none. But it's used in the whole framework for handlers for certain attributes. It's deeply macro-ified, but you need "something" in place for the <Start ...> element. The problem is that you need special handlers depending on if it's a BinaryVariableStart, StringVariableStart or so.

File: src/XML/src/FMI2/fmi2_xml_type_impl.h [hunk 18]
```diff
@@ -207,14 +199,6 @@ fmi2_xml_integer_type_props_t *fmi2_xml_parse_integer_type_properties(fmi2_xml_p
...
 extern int fmi2_check_last_elem_is_specific_type(fmi2_xml_parser_context_t *context);
 
-extern jm_named_ptr fmi2_xml_variable_type_alloc(fmi2_xml_parser_context_t* context, jm_vector(char)* name, jm_vector(char)* description, size_t size);
-
-extern void* fmi2_xml_variable_type_create(fmi2_xml_parser_context_t* context, size_t size, jm_vector(jm_named_ptr)* typeList );
-
-extern fmi2_xml_real_typedef_t* fmi2_xml_variable_type_create_real(fmi2_xml_parser_context_t* context, fmi2_xml_elm_enu_t elmID, jm_vector(jm_named_ptr)* typeList );
-
-extern fmi2_xml_integer_typedef_t* fmi2_xml_variable_type_create_integer(fmi2_xml_parser_context_t* context, fmi2_xml_elm_enu_t elmID, jm_vector(jm_named_ptr)* typeList );
-
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI2/fmi2_xml_type_impl.h:217]: Yes, so I just removed

File: src/XML/src/FMI1/fmi1_xml_type_impl.h [hunk 19]
```diff
@@ -187,14 +187,6 @@ fmi1_xml_integer_type_props_t *fmi1_xml_parse_integer_type_properties(fmi1_xml_p
...
 extern int fmi1_check_last_elem_is_specific_type(fmi1_xml_parser_context_t *context);
 
-extern jm_named_ptr fmi1_xml_variable_type_alloc(fmi1_xml_parser_context_t* context, jm_vector(char)* name, jm_vector(char)* description, size_t size);
-
-extern void* fmi1_xml_variable_type_create(fmi1_xml_parser_context_t* context, size_t size, jm_vector(jm_named_ptr)* typeList );
-
-extern fmi1_xml_real_typedef_t* fmi1_xml_variable_type_create_real(fmi1_xml_parser_context_t* context, fmi1_xml_elm_enu_t elmID, jm_vector(jm_named_ptr)* typeList );
-
-extern fmi1_xml_integer_typedef_t* fmi1_xml_variable_type_create_integer(fmi1_xml_parser_context_t* context, fmi1_xml_elm_enu_t elmID, jm_vector(jm_named_ptr)* typeList );
-
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI1/fmi1_xml_type_impl.h:197]: Yes, so I just removed

File: Test/fmi_testutil.h [hunk 20]
```diff
@@ -21,6 +25,12 @@ void fmi_testutil_build_xml_path(char* buf, size_t bufSize, const char* basePath
  */
 void fmi_testutil_enter_breakpoint();
 
+/**
+ * Parses the modelDescription.xml which is located in 'xmldir'. Expects successful parsing.
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply Test/fmi_testutil.h:29]: It's in a bit strange state right now. There are ASSERT_MSG macros which do not kill the test, but print and return 0 (or null - depending on contexts since it's a macro). We really need to clean that up. I removed the 'Expects successful parsing' part since it's not REQUIRED.

File: Config.cmake/runtime_test.cmake [hunk 21]
```diff
@@ -11,79 +11,54 @@
 #    You should have received a copy of the FMILIB_License.txt file
 #    along with this program. If not, contact Modelon AB <http://www.modelon.com>.
 
-set(RTTESTDIR ${FMILIBRARYHOME}/Test)
+set(FMIL_TEST_DIR ${FMILIBRARYHOME}/Test)
+set(CATCH_INCLUDE_DIR ${FMILIB_THIRDPARTYLIBS}/Catch2/single_include/catch2)
```
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply Config.cmake/runtime_test.cmake:15]: Right, fixed!

**Comment by filip-stenstrom** [comment]
> Looks very good!
> 
> Some overall comments not bound to any specific code segment:
> 
> 1. Comment style // vs /* */ is a bit inconsistent. Is this worthwhile fixing? Does this need to be C89 compliant?
> 2. There are quite a few TODOs, XXX, and 1 BUG in comments. Are we keeping proper track of all of these, are theses expected to be resolved within this MR?
> 3. There are a quite a few "User is responsible for freeing memory etc comments" I think it could be useful provide the how, since it might not be clear how to  do this. And even if it is a simple free(...), this may not be obvious either.

Thanks for review. Great that you managed to find some stuff even when haven't seen code before!

1. I now do // for comments in code, because that's more convenient, and /** */ for documentation. I think that's how we should do it moving forward. We're not C89 compliant any longer due to the new variable types, which require stdint.h and stdbool.h which are C99. I still try to avoid // comments in API headers though if any user would try to macroify out the new types.
2. No, I don't track them externally. Unless it clearly should be fixed in the scope of the PR, I think we should go through them when we got a minimal working base for FMI3.
3. It's always some kind of _list object, and their API has a free method. A user should never use pure `free`. I think it's more of a problem that we're not consistent or even documenting everywhere when a user needs to handle memory

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c:133]: Ah, thanks for clarifying

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c:424]: Can we add this as a comment somehow. It does make sense but makes it hard to follow the code, unless one knows what is happening inside the called functions.

File: src/XML/src/FMI3/fmi3_xml_type_impl.h [hunk 22]
```diff
@@ -168,20 +202,63 @@ typedef struct fmi3_xml_enum_typedef_props_t {
...
+    fmi3_xml_variable_type_base_t super;
+
+    jm_string mimeType;
+    size_t maxSize;
+} fmi3_xml_binary_type_props_t;
+
+typedef struct fmi3_xml_clock_type_props_t {
+    fmi3_xml_variable_type_base_t super;
+
+    fmi3_boolean_t canBeDeactivated;
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_type_impl.h:194]: Right, makes sense. Thanks for clarifying.

File: src/XML/src/FMI3/fmi3_xml_type_impl.h [hunk 23]
```diff
@@ -168,20 +202,63 @@ typedef struct fmi3_xml_enum_typedef_props_t {
...
+typedef struct fmi3_xml_string_variable_start_t {
     fmi3_xml_variable_type_base_t super;
     char start[1];
-} fmi3_xml_variable_start_string_t;
+} fmi3_xml_string_variable_start_t;
+
+typedef struct fmi3_xml_binary_variable_start_t {
+    fmi3_xml_variable_type_base_t super;
+    size_t nStart;
+    fmi3_uint8_t start[1]; /* NOTE: Can be longer than 1. Memory can be allocated outside of struct boundary. */
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_type_impl.h:230]: Right, I misread that part, thanks for clarifying. 

Thoughts on changing to fmi3_binary_t? It is the same basetype either way, but makes it more consistent.

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 24]
```diff
@@ -1268,48 +1536,200 @@ int fmi3_xml_handle_BooleanVariable(fmi3_xml_parser_context_t *context, const ch
...
+            
+            // mimeType:
+            if (hasMimeType) {
+                jm_vector(char)* mimeType = fmi3_xml_reserve_parse_buffer(context, 3, 100);
+                if (!mimeType) return -1;
+                if (fmi3_xml_set_attr_string(context, fmi3_xml_elmID_BinaryVariable, fmi_attr_id_mimeType, 0, mimeType)) {
+                    return -1;
+                }
+                if (jm_vector_get_size(char)(mimeType) == 0) {
+                    // XXX: Doesn't seem like we proprely store the empty string? Will trigger assertion
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_variable.c:1586]: Agree on the questionable use part. However, should be try to document or note it down somewhere that this edge case may not not be treated correctly?

File: src/XML/src/FMI3/fmi3_xml_variable.c [hunk 25]
```diff
@@ -1322,7 +1742,91 @@ int fmi3_xml_handle_StringVariable(fmi3_xml_parser_context_t *context, const cha
...
+ *
+ * @param hexstr The input string to scan.
+ * @param bytearr (Output arg.) A buffer which must be able to fit the scanned bytes.
+*/
+static int fmi3_xml_hexstring_to_bytearray(fmi3_xml_parser_context_t* context, const char* hexstr, uint8_t* bytearr) {
+    size_t len = strlen(hexstr);
+    if (len % 2 != 0) {
+        // Note: 2 hexadecimal characters represent 1 byte.
+        fmi3_xml_parse_error(context, "Hexadecimal string is not of even length: %s", hexstr);
+        // TODO: Or just warning and pad with trailing '0'?
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_variable.c:1767]: Seems reasonable to skip for now. I think any format that we will find again later seems good.

File: src/XML/src/FMI3/fmi3_xml_type_impl.h [hunk 26]
```diff
@@ -26,47 +26,119 @@
...
+ *   For Enums there are two different property structs since type definition
+ *   gives the list of items and variables give min and max.
  */
 
 typedef enum {
-    fmi3_xml_type_struct_enu_typedef,
+    /**
+     * A TypeDefinition. The next node is always a _props specific to the primitive type for
+     * this TypeDefinition. The default types (i.e. if a variable doesn't have any declaredType)
+     * are not represented by any _typedef - only by _props.
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_type_impl.h]: Looks good!

**Review by PeterMeisrimelModelon — APPROVED** [review]

**Comment by PeterMeisrimelModelon** [comment]
> > Looks very good!
> > Some overall comments not bound to any specific code segment:
> > 
> > 1. Comment style // vs /* */ is a bit inconsistent. Is this worthwhile fixing? Does this need to be C89 compliant?
> > 2. There are quite a few TODOs, XXX, and 1 BUG in comments. Are we keeping proper track of all of these, are theses expected to be resolved within this MR?
> > 3. There are a quite a few "User is responsible for freeing memory etc comments" I think it could be useful provide the how, since it might not be clear how to  do this. And even if it is a simple free(...), this may not be obvious either.
> 
> Thanks for review. Great that you managed to find some stuff even when haven't seen code before!
> 
> 1. I now do // for comments in code, because that's more convenient, and /** */ for documentation. I think that's how we should do it moving forward. We're not C89 compliant any longer due to the new variable types, which require stdint.h and stdbool.h which are C99. I still try to avoid // comments in API headers though if any user would try to macroify out the new types.
> 2. No, I don't track them externally. Unless it clearly should be fixed in the scope of the PR, I think we should go through them when we got a minimal working base for FMI3.
> 3. It's always some kind of _list object, and their API has a free method. A user should never use pure `free`. I think it's more of a problem that we're not consistent or even documenting everywhere when a user needs to handle memory

1. How do we stay somewhat consistent with this. Any comment is a sort of documentation. Or do you propose // for in-line comments vs /* ... */ anything going into doxygen? 
2. Seems reasonable.
3. Sounds like it should be reasonably easy to find the appropriate free method then and I suppose we'll improve documentation as we go then.

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_parser.c:434]: Well yes, so I think the confusing part is that we even mention the required attribute. If we just didn't mention it, and it was not used anywhere else, it would be clear that the check was in that other function. But I changed it a bit to make it clear.

File: src/XML/src/FMI3/fmi3_xml_type_impl.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_type_impl.h:230]: fmi3_binary_t == fmi3_byte_t* == fmi3_uint8_t*, so we'd end up with a pointer again, not an array

**Comment by filip-stenstrom** [comment]
> > > Looks very good!
> > > Some overall comments not bound to any specific code segment:
> > > 
> > > 1. Comment style // vs /* */ is a bit inconsistent. Is this worthwhile fixing? Does this need to be C89 compliant?
> > > 2. There are quite a few TODOs, XXX, and 1 BUG in comments. Are we keeping proper track of all of these, are theses expected to be resolved within this MR?
> > > 3. There are a quite a few "User is responsible for freeing memory etc comments" I think it could be useful provide the how, since it might not be clear how to  do this. And even if it is a simple free(...), this may not be obvious either.
> > 
> > 
> > Thanks for review. Great that you managed to find some stuff even when haven't seen code before!
> > 
> > 1. I now do // for comments in code, because that's more convenient, and /** */ for documentation. I think that's how we should do it moving forward. We're not C89 compliant any longer due to the new variable types, which require stdint.h and stdbool.h which are C99. I still try to avoid // comments in API headers though if any user would try to macroify out the new types.
> > 2. No, I don't track them externally. Unless it clearly should be fixed in the scope of the PR, I think we should go through them when we got a minimal working base for FMI3.
> > 3. It's always some kind of _list object, and their API has a free method. A user should never use pure `free`. I think it's more of a problem that we're not consistent or even documenting everywhere when a user needs to handle memory
> 
> 1. How do we stay somewhat consistent with this. Any comment is a sort of documentation. Or do you propose // for in-line comments vs /* ... */ anything going into doxygen?
> 2. Seems reasonable.
> 3. Sounds like it should be reasonably easy to find the appropriate free method then and I suppose we'll improve documentation as we go then.

1. /** */ for function and struct declarations, doxygen groups (which I hope we won't need), and possibly on struct members. More or less interface documentation. // everywhere else. Up for discussion of course

File: src/XML/src/FMI3/fmi3_xml_parser.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_parser.c:434]: Great, thanks for clarifying this!

File: src/XML/src/FMI3/fmi3_xml_type_impl.h (hunk 23 above)
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_type_impl.h:230]: Right, my bad.

**Review by PeterMeisrimelModelon — APPROVED** [review]

File: src/XML/src/FMI3/fmi3_xml_parser.c [hunk 27]
```diff
@@ -891,6 +891,7 @@ int fmi3_xml_set_attr_array(fmi3_xml_parser_context_t *context, fmi3_xml_elm_enu
 
     /* write all attributes to array of correct type */
     if (fmi3_xml_str_to_array_floatXX(context, str, arrPtr, arrSize, primType)) {
+        fmi3_xml_parse_attr_error(context, elmID, attrID, str);
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_parser.c:894]: @PeterMeisrimelModelon
I removed this when cleaning up error messages during review. I realized it would be still good to still have it. Reasoning is that you should give an error message that contains the context of the failure. That context is {elmID, attrID, str}, which are not passed to fmi3_xml_str_to_array_floatXX.
